central/complianceoperator/v2/report/manager/utils/results_aggregator.go

```diff
@@ -0,0 +1,206 @@
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/search"
+)
+
+const (
+	DATA_NOT_AVAILABLE = "Data Not Available"
+	NO_REMEDIATION     = "No Remediation Available"
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+type ResultsAggregator struct {
+	checkResultsDS   checkResults.DataStore
+	scanDS           scanDS.DataStore
+	profileDS        profileDS.DataStore
+	remediationDS    remediationDS.DataStore
+	benchmarkDS      benchmarksDS.DataStore
+	complianceRuleDS complianceRuleDS.DataStore
+
+	walkByQuery walkByQueryWrapper
+}
+
+func NewReportDataGenerator(
+	checkResultsDS checkResults.DataStore,
+	scanDS scanDS.DataStore,
+	profileDS profileDS.DataStore,
+	remediationDS remediationDS.DataStore,
+	benchmarksDS benchmarksDS.DataStore,
+	complianceRuleDS complianceRuleDS.DataStore,
+) *ResultsAggregator {
+	ret := &ResultsAggregator{
+		checkResultsDS:   checkResultsDS,
+		scanDS:           scanDS,
+		profileDS:        profileDS,
+		remediationDS:    remediationDS,
+		benchmarkDS:      benchmarksDS,
+		complianceRuleDS: complianceRuleDS,
+	}
+	ret.walkByQuery = ret.WalkByQuery
+	return ret
+}
+
+type checkResultWalkByQuery func(*storage.ComplianceOperatorCheckResultV2) error
+type walkByQueryWrapper func(context.Context, string, *[]*report.ResultRow, *checkStatus) checkResultWalkByQuery
+
+// GetReportData returns map of cluster id and slice of ResultRow
+func (g *ResultsAggregator) GetReportData(req *report.Request) *report.Results {
+	resultsCSV := make(map[string][]*report.ResultRow)
+	reportResults := &report.Results{}
+	for _, clusterID := range req.ClusterIDs {
+		clusterResults, clusterStatus, err := g.getReportDataForCluster(req.Ctx, req.ScanConfigID, clusterID)
+		if err != nil {
+			log.Errorf("Data not found for cluster %s", clusterID)
+			continue
+		}
+		resultsCSV[clusterID] = clusterResults
+		reportResults.TotalPass += clusterStatus.totalPass
+		reportResults.TotalFail += clusterStatus.totalFail
+		reportResults.TotalMixed += clusterStatus.totalMixed
+	}
+	reportResults.Clusters = len(req.ClusterIDs)
+	reportResults.Profiles = req.Profiles
+	reportResults.ResultCSVs = resultsCSV
+	return reportResults
+}
+
+func (g *ResultsAggregator) getReportDataForCluster(ctx context.Context, scanConfigID, clusterID string) ([]*report.ResultRow, *checkStatus, error) {
+	var ret []*report.ResultRow
+	scanConfigQuery := search.NewQueryBuilder().AddExactMatches(search.ComplianceOperatorScanConfig, scanConfigID).
+		AddExactMatches(search.ClusterID, clusterID).
+		ProtoQuery()
+	statuses := &checkStatus{
+		totalPass:  0,
+		totalFail:  0,
+		totalMixed: 0,
+	}
+	err := g.checkResultsDS.WalkByQuery(ctx, scanConfigQuery, g.walkByQuery(ctx, clusterID, &ret, statuses))
+	return ret, statuses, err
+}
+
+func (g *ResultsAggregator) WalkByQuery(ctx context.Context, clusterID string, clusterResults *[]*report.ResultRow, checkStatus *checkStatus) checkResultWalkByQuery {
```

Comment:
I'm not sure about the name for this one or the wrapper.  feels more like an aggregation function than a walk to me as it is processing the row from the results Walk to get the additional information.

Reply:
Yeah I was not very happy either. I've changed it to `AggregateResults`. Let me know what you think.

Reply:
I like that better.  Thanks